### PR TITLE
feat(ingestion/powerbi): support advance sql construct in native query parsing

### PR DESCRIPTION
**Pull request by siddiquebagwan**

### CODE REVIEW

File: metadata-ingestion/docs/sources/powerbi/powerbi_pre.md [hunk 1]
```diff
@@ -40,9 +40,17 @@ PowerBI Source supports M-Query expression for below listed PowerBI Data Sources
...
+Native SQL query parsing is supported for `Snowflake` and `Amazon Redshift`. To enable native SQL query parsing enable `native_query_parsing` and `enable_advance_lineage_sql_construct` flags in recipe.
+
+<br/>
+
+> caveat: By default `convert_lineage_urns_to_lowercase` is enabled, in-case if you have disabled it in previous ingestion execution then it may break lineage as this option generates the upstream datasets URN in lowercase.
+
+<br/>
+
+
+Use full-table-name in `from` clause. for example `operations_analytics.transformed_prod.v_unit_targets` in below native SQL query. The table `operations_analytics.transformed_prod.v_unit_targets` will be ingested as upstream table.
```
Comment (hsheth2) [inline metadata-ingestion/docs/sources/powerbi/powerbi_pre.md:52]: these docs are getting kinda unwieldy and hard to read - could you put high level stuff here + create a new section below that contains the mquery parsing details + code blocks and things

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/native_sql_parser.py [hunk 2]
```diff
@@ -41,7 +44,48 @@ def get_tables(native_query: str) -> List[str]:
         logger.debug(f"Type={type(tokens[from_index])}")
         if isinstance(tokens[from_index], sqlparse.sql.Identifier):
             # Split on as keyword and collect the table name from 0th position. strip any spaces
-            tables.append(tokens[from_index].value.split("as")[0].strip())
+            table: str = tokens[from_index].value.split("as")[0].strip()
+
+            if len(table.split(".")) == 1:
+                table = f"{default_schema}.{table}"
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/native_sql_parser.py:50]: this definitely won't create valid urns?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/native_sql_parser.py [hunk 3]
```diff
@@ -41,7 +44,48 @@ def get_tables(native_query: str) -> List[str]:
...
+
+
+def _advance_sql_query_parsing(native_query: str, default_schema: str) -> List[str]:
+    """
+    This function uses sqllineage package to parse native sql.
+    It supports normal as well as advance sql query construct.
+    """
+    tables: List[str] = []
+
+    parser = LineageRunner(native_query)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/native_sql_parser.py:66]: we have logic around LineageRunner in a ton of places in our codebase

does it make sense to refactor those? I'm also working on some sql parsing stuff, so having a cleanly defined interface without all this copy-pasted code would make my life a lot easier

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py [hunk 4]
```diff
@@ -352,7 +362,7 @@ def resolve_to_data_platform_table_list(self) -> List[DataPlatformTable]:
                 continue
 
             table_full_name_creator: AbstractDataPlatformTableCreator = (
-                supported_resolver.get_table_full_name_creator()()
+                supported_resolver.get_table_full_name_creator()(config=self.config)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py:365]: what does this do here?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/native_sql_parser.py (hunk 3 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/native_sql_parser.py:66]: Agree, check this thread as well https://acryldata.slack.com/archives/C02ENS87APK/p1685715808871169?thread_ts=1685715193.359649&cid=C02ENS87APK

So Shall I use that interface in this connector initially and then have a separate PR for all of the connectors?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/native_sql_parser.py (hunk 2 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/native_sql_parser.py:50]: It is not creating URN. It is only returning the table name in the form of <schema-name>.<table-name>. it is old code and existing test cases are working. The difference is; added `default_schema` as an argument.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py (hunk 4 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py:365]: We need config `enable_advance_lineage_sql_construct` in full-table-name creator to decide whether to go for advanced native query parsing.
Please check config.py: https://github.com/datahub-project/datahub/pull/8251/files#diff-f32896b0964b087e35e8a171a0b941c6aa0e53a5e5d42cb243b7fdd6e4eab3bcR384

**Comment by siddiquebagwan-gslab** [comment]
It is duplicate of https://github.com/datahub-project/datahub/pull/8592

**Pull request closed without merging**